lucet-idl/src/witx/parser.rs

```diff
@@ -0,0 +1,557 @@
+use super::sexpr::SExpr;
+use super::Location;
+
+///! Parser turns s-expressions into unvalidated syntax constructs.
+///! conventions:
+///! `Type::starts_parsing(s-expr) -> bool` is for look-ahead: we use
+///! this predicate to combine parsers for different `Type`s where both
+///! alternatives are accepted.
+///! `Type::parse(sexpr: &SExpr) -> Result<Self, ParseError>` takes a single
+///! s-expression and parses it into a `Self`.
+///! for parsers that take a subset of a vector s-expression, the signature
+///! `Type::parse(sexprs: &[SExpr], location: Location) -> Result<Self, ParseError>`
+///! has an additional `Location` argument, which should point to the parent SExpr::Vec.
+///! This is used for error reporting in case the slice doesn't have the number of elements
+///! expected.
+
+#[derive(Debug, Fail)]
+#[fail(display = "{} at {:?}", _0, _1)]
+pub struct ParseError {
+    pub message: String,
+    pub location: Location,
+}
+
+macro_rules! parse_err {
+    ($loc:expr, $msg:expr) => {
+        ParseError { message: $msg.to_string(), location: $loc.clone() }
+    };
+    ($loc:expr, $fmt:expr, $( $arg:expr ),+ ) => {
+        ParseError { message: format!($fmt, $( $arg ),+), location: $loc.clone() }
+    };
+}
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct IdentSyntax {
+    pub name: String,
+    pub location: Location,
+}
+
+macro_rules! id {
+    ($s:expr, $loc: expr) => {
+        IdentSyntax {
+            name: $s.to_string(),
+            location: $loc.clone(),
+        }
+    };
+}
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub enum BuiltinType {
+    String,
+    Data,
```

Comment:
Its a byte slice. I'd expect for it to eventaully be changed to use the Array constructor and U8, but at the moment its special cased in interface-types because they havent figured out arrays yet, and neither have we really